opencog/atomspace/AtomTable.h

```diff
@@ -271,18 +312,31 @@ class AtomTable
      */
     template <typename OutputIterator> OutputIterator
     getHandlesByTargetType(OutputIterator result,
-                             Type type,
+                             Type resultAtomType,
                              Type targetType,
-                             bool subclass,
+                             bool resultAtomSubclass,
                              bool targetSubclass) const
     {
         std::lock_guard<std::recursive_mutex> lck(_mtx);
-        return std::copy_if(targetTypeIndex.begin(targetType, targetSubclass),
-                            targetTypeIndex.end(),
-                            result,
-             [&](Handle h)->bool{
-                 return isDefined(h) and h->isType(type, subclass);
+
+        // We're going to use this one handle so we don't construct/destruct
+        // googles of temporaries in the process of doing the copies.
+        Handle temp_handle;
+        std::for_each(targetTypeIndex.begin(targetType, targetSubclass),
+                      targetTypeIndex.end(),
+            [&](UUID uuid)->void {
+                if (UNDEFINED_UUID == uuid) return;
+                
+                // Since we resolve the pointer below, we call the safe
+                // version of set_uuid which clears the AtomPtr reference.
+                temp_handle.set_uuid(uuid);
+
+                // If the predicate function returns true, then copy this
+                // handle into the result. The -> resolves the AtomPtr.
+                if (temp_handle->isType(resultAtomType, resultAtomSubclass))
+                    *result = temp_handle;
              });
+        return result;
```

Comment:
all of these different getHandlesByWhatever functions are semi-deprecated. The correct fix is not to improve them, but figure out who is using them, and why, and try to understand and eliminate those usages.